plaid/src/apis/npm/npm_web_client.rs

```diff
@@ -761,6 +778,69 @@ impl Npm {
             page_num += 1;
         }
     }
+
+    /// Return a JSON-encoded list of packages that a granular token can publish
+    pub async fn get_token_publish_scope(
+        &self,
+        params: &str,
+        module: &str,
+    ) -> Result<String, ApiError> {
+        let params: GetTokenPublishScopeParams =
+            serde_json::from_str(params).map_err(|_| ApiError::BadRequest)?;
+        let token_id = self.validate_token_id(&params.token_id)?;
+
+        info!("Retrieving publish scope for token [{token_id}] on behalf of [{module}]");
+
+        self.login()
+            .await
+            .map_err(|_| ApiError::NpmError(NpmError::LoginFlowError))?;
+        let response = self
+            .client
+            .get(format!(
+                "{}/settings/{}/tokens/granular-access-tokens/{}",
+                NPMJS_COM_URL, self.config.username, token_id
+            ))
+            .header("X-Spiferack", "1") // to get JSON instead of HTML
+            .send()
+            .await
+            .map_err(|_| ApiError::NpmError(NpmError::FailedToGetTokenPublishScope))?;
+
+        // Information on the token is returned in an object under "tokenDetails"
+        let token_details = serde_json::from_value::<GranularTokenDetails>(
+            response
+                .json::<Value>()
+                .await
+                .map_err(|_| ApiError::NpmError(NpmError::FailedToGetTokenPublishScope))?
+                .get("tokenDetails")
+                .ok_or(ApiError::NpmError(NpmError::FailedToGetTokenPublishScope))?
+                .clone(),
+        )
+        .map_err(|_| ApiError::NpmError(NpmError::FailedToGetTokenPublishScope))?;
+
+        // Check a few things about the token before returning its publish scope
+
+        if token_details.expired
```

Comment:
This is counter intuitive to the caller, that just because a token is expired, the scopes disappear. I would wonder if we should build a more generic API to just return granular token details and work with specifics on the module side. Thoughts?

Reply:
Good idea and it would keep things easier on the runtime's side. So I will make a `get_token_details` API and return a `GranularTokenDetails` struct.
As for expired tokens, admittedly I don't know what happens when an NPM token expires. I have just created one which expires tomorrow so I'll wait and see.